lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,154 @@
+#define DEBUG_TYPE "opaque-archetype-specializer"
+
+#include "swift/SIL/SILFunction.h"
+#include "swift/SIL/TypeSubstCloner.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+
+
+using namespace swift;
+
+namespace {
+class OpaqueSpecializerCloner
+    : public TypeSubstCloner<OpaqueSpecializerCloner,
+                      SILOptFunctionBuilder /*unused*/> {
+
+  using SuperTy =
+      TypeSubstCloner<OpaqueSpecializerCloner, SILOptFunctionBuilder>;
+
+  SILBasicBlock *entryBlock;
+  SILBasicBlock *cloneFromBlock;
+
+public:
+  friend class SILCloner<OpaqueSpecializerCloner>;
+  OpaqueSpecializerCloner(SubstitutionMap opaqueArchetypeSubs, SILFunction &fun)
+      : SuperTy(fun, fun, opaqueArchetypeSubs) {
+    entryBlock = fun.getEntryBlock();
+    cloneFromBlock = entryBlock->split(entryBlock->begin());
+  }
+
+  void clone();
+
+protected:
+  void insertOpaqueToConcreteAddressCasts(SILInstruction *orig,
+                                          SILInstruction *cloned);
+
+  void postProcess(SILInstruction *orig, SILInstruction *cloned) {
+    SILClonerWithScopes<OpaqueSpecializerCloner>::postProcess(orig, cloned);
+    insertOpaqueToConcreteAddressCasts(orig, cloned);
+  }
+
+  void visitTerminator(SILBasicBlock *BB) { visit(BB->getTerminator()); }
+
+  SILType remapType(SILType Ty) {
+    SILType &Sty = TypeCache[Ty];
+    if (!Sty) {
+      // Apply the opaque types substitution.
+      ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+      Sty = Ty.subst(Original.getModule(), SubsMap)
+                .subst(Original.getModule(), replacer, replacer,
+                       CanGenericSignature(), true);
+    }
+    return Sty;
+  }
+
+  CanType remapASTType(CanType ty) {
+    // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapASTType(ty)
+        .subst(replacer, replacer,
+               SubstFlags::SubstituteOpaqueArchetypes |
+                   SubstFlags::AllowLoweredTypes)
+        ->getCanonicalType();
+  }
+
+  ProtocolConformanceRef remapConformance(Type type,
+                                          ProtocolConformanceRef conf) {
+    // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapConformance(type, conf)
+        .subst(type, replacer, replacer,
+               SubstFlags::SubstituteOpaqueArchetypes |
+                   SubstFlags::AllowLoweredTypes);
+  }
+
+  SubstitutionMap remapSubstitutionMap(SubstitutionMap Subs) {
+    // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapSubstitutionMap(Subs).subst(
+        replacer, replacer,
+        SubstFlags::SubstituteOpaqueArchetypes | SubstFlags::AllowLoweredTypes);
```

Comment:
This formatting is not consistent with the previous lines

Reply:
clang-format does that.